sources/web/datalab/static/datalab.js

```diff
@@ -277,6 +277,166 @@ function initializePage(dialog, saveFn) {
+  let inputDiv = cell.cell_type === 'code' ? 'div.input' : 'div.inner_cell';
+  let isHidden = typeof(hide) !== 'undefined' ? !hide : cell.element.find(inputDiv)[0].style.display === 'none';
+  collapseSpan = cell.element.find('span.collapse-cell')[0];
+  if (isHidden) {
+    cell.element.find(inputDiv).show();
+    // for code cells, also show the output div
+    if (cell.cell_type == 'code')
+      cell.element.find('div.output_wrapper').show();
+    cell.element.find('div.cellPlaceholder').hide();
+    collapseSpan.classList.add(UNCOLLAPSE_BUTTON_CLASS);
+    collapseSpan.classList.remove(COLLAPSE_BUTTON_CLASS);
+  } else {
+    cell.element.find(inputDiv).hide();
+    // for code cells, also hide the output div
+    if (cell.cell_type == 'code')
+      cell.element.find('div.output_wrapper').hide();
+    cell.element.find('div.cellPlaceholder').show();
+    collapseSpan.classList.remove(UNCOLLAPSE_BUTTON_CLASS);
+    collapseSpan.classList.add(COLLAPSE_BUTTON_CLASS);
+  }
+  cell.metadata[CELL_METADATA_COLLAPSED] = !isHidden;
+}
+
+/**
+ * Collapse output of a cell
+ */
+function toggleCollapseCellOutput(cell, hide) {
+  // the cell's output is currently collapsed if its output div is hidden
+  let isHidden = typeof(hide) !== 'undefined' ? !hide : cell.element[0].getElementsByClassName('output')[0].style.display === 'none';
+  collapseSpan = cell.element.find('span.collapse-output')[0];
+  if (isHidden) {
+    cell.expand_output();
+    collapseSpan.classList.add(COLLAPSE_OUTPUT_BUTTON_CLASS);
+    collapseSpan.classList.remove(UNCOLLAPSE_OUTPUT_BUTTON_CLASS);
+  } else {
+    cell.collapse_output();
+    collapseSpan.classList.remove(COLLAPSE_OUTPUT_BUTTON_CLASS);
+    collapseSpan.classList.add(UNCOLLAPSE_OUTPUT_BUTTON_CLASS);
+  }
+  cell.metadata[CELL_METADATA_OUTPUT_COLLAPSED] = !isHidden;
+}
+
+/**
+ * Create an HTML button for the cell minitoolbar and return it
+ */
+function createCellMiniToolbarButton(classNames, title, callback) {
+  let buttonDiv = document.createElement('button');
+  buttonDiv.className = 'btn btn-default';
+  buttonDiv.innerHTML = '<span class="' + classNames + '"></span>';
+  buttonDiv.title = title;
+  buttonDiv.addEventListener('click', callback);
+  return buttonDiv;
+}
+
+/**
+ * Patch the cell's element to add the hoverable minitoolbar
+ */
+function addCellMiniToolbar() {
+  Jupyter.notebook.get_cells().forEach(function(cell) {
+
+    let hoverableDiv = document.createElement('div');
+    hoverableDiv.className = 'hoverable';
+
+    // only for code cells, add collapse-output and clear buttons
+    if (cell.cell_type === 'code') {
+      // collapse output button
+      let hideButton = createCellMiniToolbarButton(
+        'collapse-output fa ' + COLLAPSE_OUTPUT_BUTTON_CLASS,
+        'Collapse/Expand output',
+        function() {
+          toggleCollapseCellOutput(cell);
+        }
+      );
+      cell.element.find('div.output_collapsed')[0].addEventListener('click', function() {
```

Comment:
Multiple cell outputs are possible (I think multiple calls of IPython.core.display.*() will do it). Can we make sure all of them are added the minitool bar?